### PR TITLE
Airways page improvements

### LINKED ISSUE #1550: [BUG] Airways do not work as expected with a long route
**Mod Version**
master - A32NX-master.zip
built: 2020-10-26T03:15:17+00:00
ref: refs/heads/master
sha: 8ba938f0bfa1c07e1f60ae34c41385507fe4a6e0
actor: Benjozork
event_name: push

Using Navigraph data

**Describe the bug**
The new airway support does not appear to work as expected after 2-3 pages of airways. This is a combination of airways between waypoints missing in the full flightplan and then the airways page clearing upon entry of first airway/waypoint combination on subsequent pages.

**To Reproduce**
1. Start new flight cold and dark
2. Initialise flight between LFBO and OMDB
3. Insert following flight plan

`LFBO/32R AMOL5B AMOLO UT24 FJR UL127 SODRI UM728 DOKAR DCT TAQ DCT PEMAR DCT EKMOT DCT DIVKU DCT ENOXA Y2 PAPIZ DCT PINDO UP14 RIKSO UN127 IMR UN135 VESAR L619 NIKAS R785 KTN A21 TAN G202 PUSTO M203 ILMAP UP975 ULDUR UM688 SIDAD R784 TATLA L223 VUTEB VUTE3C OMDB/ULDOT.R30L`

**Expected behavior**
It should be possible to add a flight of any length and number of airways via the new airway support.

**Actual behavior**
1. After 3 pages of airways, new waypoints do not have airways between them in flight-plan page
2. After another page, after entering first airway and waypoint, the airway and waypoint will disappear off the airway page, but the waypoint and its intermediaries will be added to the flight plan (with airways between them included)
3. This can then be repeated for all subsequent airways (by going to last waypoint in the flight plan once at a time)

Was this working before/when did the issue start occurring?
This started to occur since the new airway support was added yesterday. The functionality is brilliant, especially for shorter flights, but with a couple of corner cases in this longer example.

Is this a problem in the vanilla unmodded game?
No as is related to the new airway support that was added

Discord username (if different from GitHub): sawlo

### PR DESCRIPTION
**Pull request by santii90**
Fixes #1550 

## Summary of Changes

- This add the correct behavior while adding them, having a temporary plan and not adding al they waypoints as soon as the last airway WP is added.
- Fixed a weird route that was not working, and I expect this to fix other similar routes. Route for reference: VGHS/32 DCT ADMIL G463 BGO L507 IGONI DCT VTBS
- Minor performance improvements.
- I disable the DIRECT TO capability and the departure being show as an initial airway/waypoint because I couldn't found references to that on the manual.

## Screenshots (if necessary)

airwaysUI4
image

## References

Changed reference to the Honeywell Pegasus Step 1A (Rev 0) and updated colors.
For the DIRECT TO behavior that I disabled I got references of that from a real life a320 pilot.

## Additional context

Discord username (if different from GitHub): santii#9289

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by mtjones83** [comment]
Name: mtjones83 (Discord MikeJ#0833)
Date of testing: 27/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1544
Object of test: Validate airways page improvements

Testing technique: 

1. Spawned on runway at EGGP (engines running)
2. Followed 'normal' process for MCDU setup: INIT A & B pages
3. Followed flight plan programming process: F-PLN page
4. Programmed the following route, incl. runways, SIDs & STARs
image
5. Stepped through flight plan on ND plan mode to validate.


Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Works well. Performance of adding airways/waypoints feels much improved. Accurate colours also good. Also fixes issue where waypoints along an airway were not loaded into the flight plan; this now works as expected. Big improvement.

 
image
image
image

**Comment by Benjozork** [comment]
@mtjones83 thanks for your report, but remember that this is not marked with the "Ready to Test" label.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js [hunk 1]
```diff
@@ -21,17 +23,24 @@ class A320_Neo_CDU_AirwaysFromWaypointPage {
...
+            if (r[0] != "" && r[1] != "") {
+                subRows[idx] = ["VIA", "TO"];
+            }
+        });
+        mcdu._titleElement.innerHTML = `<span><span>AIRWAYS</span> <span class='s-text'>FROM </span><span class='green'>${waypoint.ident}</span></span>`;
         let showInput = false;
         const departureWaypoints = mcdu.flightPlanManager.getDepartureWaypoints();
         const routeWaypoints = mcdu.flightPlanManager.getEnRouteWaypoints();
-        offset = routeWaypoints.indexOf(waypoint) >= 0 ? routeWaypoints.indexOf(waypoint) + (departureWaypoints.length ? 1 : 0) + 1 : offset;
+        // offset = routeWaypoints.indexOf(waypoint) >= 0 ? routeWaypoints.indexOf(waypoint) + (departureWaypoints.length ? 1 : 0) + 1 : offset;
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js:35]: Can this not be done using the "[color>blue" pattern I see around h ere ?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js [hunk 2]
```diff
@@ -45,79 +54,97 @@ class A320_Neo_CDU_AirwaysFromWaypointPage {
...
                             });
                         }
                     };
-                    if (rows[i + 1]) {
-                        rows[i + 1] = ["-----"];
-                    }
+                    // May be useful for intersecions.
+                    // if (rows[i + 1]) {
+                    //     rows[i + 1] = ["-----"];
+                    // }
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js:93]: Please remove commented code throughout your changes. Thanks !

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js [hunk 3]
```diff
@@ -45,79 +54,97 @@ class A320_Neo_CDU_AirwaysFromWaypointPage {
...
             rows[3],
-            [""],
+            subRows[4],
             rows[4],
             rowBottomLabel,
             rowBottomLine
         ]);
     }
-    static _GetAllRows(fmc) {
+    static _GetAllRows(mcdu,currentWP) {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js:113]:
```suggestion
    static _GetAllRows(mcdu, currentWP) {
```

**Review by Benjozork — CHANGES_REQUESTED** [review]
Thank you for your PR. Minor things here and there

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js (hunk 1 above)
Comment (santii90) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js:35]: I tried to do it with that approach but I was not able to have multiples color configuration on the same string for the title.
I'm probably missing something on the approach, if you have any reference please let me know.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js (hunk 2 above)
Comment (santii90) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js:93]: Forgot to remove, thanks.

**Comment by mtjones83** [comment]
> @mtjones83 thanks for your report, but remember that this is not marked with the "Ready to Test" label.

of course; apologies! #TooKeen ;-)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js (hunk 1 above)
Comment (Benjozork) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AirwaysFromWaypointPage.js]: We'll probably add that possibility at a later time, good for now

**Review by Benjozork — APPROVED** [review]

**Comment by Benjozork** [comment]
cc. maintainers: we can take @mtjones83's test into account since no major changes were done

**Comment by beheh** [comment]
The labels on the bottom buttons (TMPY ERASE / TMPY INSERT) in the first screenshot do not match the real reference on the left [here at 6m59s](https://www.youtube.com/watch?v=twwUjHXdNVU#t=6m59s).

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -75,6 +75,7 @@
 1. [LIGHTS] Added logo lights only turn on while on ground or flaps are extented - @linkeleven (linkeleven#2557)
 1. [CDU] Add basic airways support - @santii90 (Santiago Vazquez), @Lessar277 (Lessar27)
```
Comment (dga711) [inline .github/CHANGELOG.md:76]: missing credits for code donation from WorkingTitle

**Comment by santii90** [comment]
> The labels on the bottom buttons (TMPY ERASE / TMPY INSERT) in the second screenshot do not match the real reference on the left [here at 6m59s](https://www.youtube.com/watch?v=twwUjHXdNVU#t=6m59s).
> 
> From the same reference it also looks like the brackets `[ ]` should be a bit wider than on the screenshot.

Thanks, I did a small fix for the TMPY labels, now it's 100% the same as the Honeywell. For the brackets I need to do some major changes because the way Asobo uses the spaces, I'm going to wait for a general fix so that can be addressed on more pages.

File: .github/CHANGELOG.md (hunk 4 above)
Comment (Benjozork) [reply .github/CHANGELOG.md]: Sorry, I was going to ensure this was done but forgot.

**Comment by santii90** [comment]
While doing some testing flight I came up with an minor issue on the UI when selecting a departure and starting to add airways from the waypoint that link the departure to the airway.
I recommend doing some testing on this scenario and also when not selecting a departure or starting to add airways after some directs from the departure.

**Review by Carter633 — APPROVED** [review]
Name: British Avgeek YT -- Carter633 #1436
Date of testing: Current date 28/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1544 
Object of test: Check further improvements to Airways page

Testing technique: Full flight from C&D - EGKK to EGPH
Steps you've made to test the Pull Request:
1. Spawned C&D EGKK on VATSIM
2. Input of full flight plan using only the airway page after SID
3. Then input STAR from airway - worked fine
4. Checked on PLAN mode on ND
5. All worked as expected - flew the planned flight

Flight notes
ROUTE planned using SIMBRIEF:
LAM6M LAM L10 BPK/N0457F380 UN601 INPIP1E

Startup: As expected
Taxi/Takeoff : As expected
Cruise: As expected
Descent/Landing : As expected
Shutdown: N/A

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Worked as described - further improvements to the previous PR I tested a few days ago with numerous routes etc. I'll run another test with a longer route to see how it behaves with this PR (as I did the other day) - Currently no issues found

**Comment by danbw9** [comment]
Name: danbw9#7028
Date of testing: Current date 28/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1544
Object of test: Checking further improvements to Airways page

Testing technique: Try multiple scenarios to check functionality and any issues.
Steps you've made to test the Pull Request:

C&D Spawn. Input 3 different types of routes (VOR to VOR, short flight with some DCTs and low and high airways and long flight with 58 legs). 
Cross checked with flight plan route and also via PLAN mode on ND to check accuracy. 
Completed full flight from EGLL to EKCH (RTE starting with BPK) tried doing lateral revisions on the ground and in flight and it handled it fine.

Flight notes
ROUTES:
- TLA DCT HON DCT LAM
- BPK Q295 SOMVA DCT MAVAS DCT NIRDU DCT DOSUR P729 TUDLO
- TIPSA G270 AGLEK A113 TOMGO A356 DIKAM/K0845F340 A356 KZO B710 TIGTA R227 AKT G96 AGALA/K0831F360 G96 MKL A955 LURIS/N0450F360 N37 SARPI UM10 TIVUG UL852 MODAU UT643 ATPIX

Startup: Everything normal. Adjusted to the change where DCT to waypoints being done via AIRWAYS page is not possible but via the 'NEXT waypoint' function with LS keys on the left in F PLAN page. 

Taxi/Takeoff : Everything normal

Cruise: Everything normal

Descent/Landing : Everything normal

Shutdown: N/A

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: LGTM - its coming along very nicely. Thank you @santii90. This fixes the issue I had with a 58 leg route where the inputs would disappear. Performance was very good as well. Ready to merge from my side

**Comment by Carter633** [comment]
Name: British Avgeek YT -- Carter633 #1436
Date of testing: Current date 28/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1544
Object of test: Check further improvements to Airways page - TEST 2 of 2 from me for this PR.

Testing technique: Full flight from C&D - EGKK to LGIR on VATSIM
Steps you've made to test the Pull Request:

1. Spawned C&D EGKK on VATSIM
2. Input of full flight plan using only the airway page after SID
3. Then input STAR from airway - worked fine
4. Checked on PLAN mode on ND

Flight notes
ROUTE planned using SIMBRIEF:
ADMA2X ADMAG Y312 DVR UL9 KONAN UL607 REMBA DCT LIRSU DCT NONKO UL607 RIDSU UZ738 RIDAR DCT VAROB DCT ABISO/N0448F370 DCT NEMEK DCT VRANA DCT PETAK DCT PINDO UL607 GIVIS T514 XAVIS XAVI1A

Startup: Was a little confusing at first with the right hand side brackets having disappeared - all was fine until I got to the DCT --- DCT --- DCT --- part in the route above...
...so I entered the direct waypoints in the right hand side of the airways page on the MCDU anyway (clicking the button below the last airway/waypoint line) and it handled it fine though didn't list them on the airways tab as it did on the previous airway page (need an IRL pilot to tell me if that's correct or not). Either way I was able to put in the entire route on the airway tab anyway very quickly
Taxi/Takeoff : As expected
Cruise: As expected
Descent/Landing : As expected
Shutdown: N/A

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Worked as described and as above for notes from test 2. Very different to the first test and significantly longer for a route too but all looks good though, I'd like clarification on the noted point above if I've explained it well enough

**Comment by santii90** [comment]
> Name: British Avgeek YT -- Carter633 #1436
> Date of testing: Current date 28/10/2020
> Version of the sim: 1.9.5.0
> PR Tested: #1544
> Object of test: Check further improvements to Airways page - TEST 2 of 2 from me for this PR.
> 
> Testing technique: Full flight from C&D - EGKK to LGIR on VATSIM
> Steps you've made to test the Pull Request:
> 
> 1. Spawned C&D EGKK on VATSIM
> 2. Input of full flight plan using only the airway page after SID
> 3. Then input STAR from airway - worked fine
> 4. Checked on PLAN mode on ND
> 
> Flight notes
> ROUTE planned using SIMBRIEF:
> ADMA2X ADMAG Y312 DVR UL9 KONAN UL607 REMBA DCT LIRSU DCT NONKO UL607 RIDSU UZ738 RIDAR DCT VAROB DCT ABISO/N0448F370 DCT NEMEK DCT VRANA DCT PETAK DCT PINDO UL607 GIVIS T514 XAVIS XAVI1A
> 
> Startup: Was a little confusing at first with the right hand side brackets having disappeared - all was fine until I got to the DCT --- DCT --- DCT --- part in the route above...
> ...so I entered the direct waypoints in the right hand side of the airways page on the MCDU anyway (clicking the button below the last airway/waypoint line) and it handled it fine though didn't list them on the airways tab as it did on the previous airway page (need an IRL pilot to tell me if that's correct or not). Either way I was able to put in the entire route on the airway tab anyway very quickly
> Taxi/Takeoff : As expected
> Cruise: As expected
> Descent/Landing : As expected
> Shutdown: N/A
> 
> Issues (If Any): None
> Overall Rating: Good
> Conclusions of the testing: Worked as described and as above for notes from test 2. Very different to the first test and significantly longer for a route too but all looks good though, I'd like clarification on the noted point above if I've explained it well enough

Hello and thanks for the testing. DIRECT TO waypoint on the airways is a behavior that does not exist on the real airplane based on feedback from pilots. It's not going to be displayed on the page. And you being able to add them is a left over on the code that  I plan to totally remove on the next iteration. If QA prefer to have that removed now I can do that also.

**Comment by Carter633** [comment]
Thanks for getting back - I figured as they weren't waypoints they shouldn't be inputted into the airway page but wanted to be sure! All seems good either way - it's behaving as expected through the flight now and I've been given some additional direct to by ATC and it's not thrown any of it out yet.

**Pull request merged**